File: prep/implementation/patterns_reference_implementation/topological_sort/TopologicalSort.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class TopologicalSort {
  static interface Graph {
    void addEdge(int from, int to);
    List<Integer> getAdjacentVertices(int v);
    int indegree(int v);
    int size();
  }

  static class AdjacencySet implements Graph {
    static class Vertex {
      Set<Integer> vs;

      Vertex() { this.vs = new HashSet<>(); }

      void addEdge(int v) { this.vs.add(v); }

      List<Integer> getAdjacentVertices() { return new ArrayList<>(this.vs); }

      boolean hasEdge(int v) { return this.vs.contains(v); }
    }

    private List<Vertex> vertices;

    AdjacencySet(int size) {
      this.vertices = new ArrayList<>(size);
      for (int i = 0; i < size; i++) {
        this.vertices.add(new Vertex());
      }
    }

    @Override
    public void addEdge(int from, int to) {
      if (from < 0 || from >= size() || to < 0 || to >= size()) {
        throw new IllegalArgumentException("invalid vertices");
      }

      this.vertices.get(from).addEdge(to);
    }

    @Override
    public List<Integer> getAdjacentVertices(int v) {
      if (v < 0 || v >= size()) {
        throw new IllegalArgumentException("invalid vertex");
      }

      return this.vertices.get(v).getAdjacentVertices();
    }

    @Override
    public int indegree(int v) {
      if (v < 0 || v >= size()) {
        throw new IllegalArgumentException("invalid vertex");
      }

      int indeg = 0;
      for (int i = 0; i < size(); i++) {
        if (this.vertices.get(i).hasEdge(v)) {
          indeg++;
        }
      }

      return indeg;
    }

    @Override
    public int size() {
      return this.vertices.size();
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        Graph g = new AdjacencySet(n);

        int m = in.nextInt();
        for (int i = 0; i < m; i++) {
          int from = in.nextInt();
          int to = in.nextInt();

          g.addEdge(from, to);
        }

        var ordering = sort(g);
        for (int o : ordering) {
          System.out.printf("%d ", o);
        }
        System.out.println();
      }
    }
  }

  // O(V + E) / O(V + E)
  public static List<Integer> sort(Graph g) {
    Queue<Integer> q = new ArrayDeque<>();

    Map<Integer, Integer> indegree = new HashMap<>();
    for (int i = 0; i < g.size(); i++) {
      int d = g.indegree(i);
      indegree.put(i, d);

      if (d == 0) {
        q.add(i);
      }
    }

    List<Integer> ordering = new ArrayList<>();
    while (!q.isEmpty()) {
      int v = q.poll();

      ordering.add(v);

      for (int neighbour : g.getAdjacentVertices(v)) {
        indegree.put(neighbour, indegree.get(neighbour) - 1);
        if (indegree.get(neighbour) == 0) {
          q.add(neighbour);
        }
      }
    }

    if (ordering.size() != g.size()) {
      throw new IllegalStateException("cycle detected");
    }

    return ordering;
  }
}

File: lc/coreprep/implementation/main_iterations/iteration_cpp/topological_sort.cpp
#include <iostream>
#include <queue>
#include <set>
#include <unordered_map>
#include <vector>

using namespace std;

struct Vertex {
  set<int> vs;

  void add_edge(int v) { this->vs.insert(v); }

  const vector<int> get_adjacent_vertices() const {
    vector<int> neighbours;

    for (int v : this->vs) {
      neighbours.push_back(v);
    }

    return neighbours;
  }

  bool has_edge(int v) { return this->vs.find(v) != this->vs.end(); }
};

class Graph {
private:
  int sz;
  vector<Vertex> vertices;

public:
  Graph(int sz) : sz(sz) {
    for (int i = 0; i < sz; i++) {
      Vertex v;
      this->vertices.emplace_back(v);
    }
  }

  void add_edge(int from, int to) {
    if (from < 0 || from >= this->sz || to < 0 || to >= this->sz) {
      throw "illegal vertex";
    }

    this->vertices[from].add_edge(to);
  }

  const vector<int> get_adjacent_vertices(int v) const {
    if (v < 0 || v >= this->sz) {
      throw "illegal vertex";
    }

    return this->vertices[v].get_adjacent_vertices();
  }

  int size() const { return this->sz; }

  int indegree(int v) {
    if (v < 0 || v >= this->sz) {
      throw "illegal vertex";
    }

    int indegree = 0;
    for (int i = 0; i < size(); i++) {
      if (this->vertices[i].has_edge(v)) {
        indegree++;
      }
    }

    return indegree;
  }
};

void topological_sort(Graph &g) {
  queue<int> q;

  unordered_map<int, int> indeg;

  for (int i = 0; i < g.size(); i++) {
    int d = g.indegree(i);
    indeg[i] = d;

    if (d == 0) {
      q.push(i);
    }
  }

  vector<int> ordering;

  while (!q.empty()) {
    int v = q.front();
    q.pop();

    ordering.push_back(v);

    for (int neighbour : g.get_adjacent_vertices(v)) {
      indeg[neighbour]--;

      if (indeg[neighbour] == 0) {
        q.push(neighbour);
      }
    }
  }

  if (ordering.size() != g.size()) {
    throw "cycle detected";
  }

  for (int v : ordering) {
    cout << v << " ";
  }
  cout << "\n";
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, m, from, to;
  cin >> n >> m;

  Graph g(n);
  for (int i = 0; i < m; i++) {
    cin >> from >> to;
    g.add_edge(from, to);
  }

  topological_sort(g);

  return 0;
}